### ISSUE TITLE
problem text is spoken repeatedly, two times with each instance, until the problem is resolved

### ISSUE THREAD

**Issue by mehgcap** [microsoft/vscode#157100]
Issue Type: <b>Bug</b>

When f8 is used to find the next problem, my screen reader speaks the problem, waits about two seconds, then speaks it again. The major problem, though, is that this continues to happen. Each time I write another statement, and often when I just arrow around my code, the problem text is spoken twice. The quickest way I've found to deal with the constant interruptions is by fixing the problem, but that's not always desired, or even possible. I might know a function doesn't yet return the type it says it should, but I'm still writing the code that will get me to the correct return type. As I type/review the code, I'm constantly reminded that the return type doesn't match the signature, just because I hit f8 once to check the code I have so far.

I'm using NVDA, the latest public version (2022.2). When this problem happens, the problem text is brailled as well as being spoken. This is even more disruptive; if speech were the worst of it, at least I could still review code in braille until NVDA shuts up about the problem. I can't emphasize enough how annoying this can get.

VS Code version: Code 1.69.2 (3b889b090b5ad5793f524b5d1d39fda662b96a2a, 2022-07-18T16:12:52.460Z)
OS version: Windows_NT x64 10.0.19044
Restricted Mode: No
Remote OS version: Linux x64 4.19.0-17-amd64
Remote OS version: Linux x64 4.19.0-17-amd64
Remote OS version: Linux x64 4.19.0-17-amd64

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|AMD Ryzen 5 3600 6-Core Processor               (12 x 3600)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.95GB (8.02GB free)|
|Process Argv|--crash-reporter-id eed1c6d2-9716-48dc-a584-f8fbc80a0901|
|Screen Reader|yes|
|VM|24%|

|Item|Value|
|---|---|
|Remote|SSH: git.autodist.com|
|OS|Linux x64 4.19.0-17-amd64|
|CPUs|Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (2 x 2494)|
|Memory (System)|7.79GB (0.32GB free)|
|VM|100%|

|Item|Value|
|---|---|
|Remote|SSH: server35.autodist.com|
|OS|Linux x64 4.19.0-17-amd64|
|CPUs|Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (2 x 2494)|
|Memory (System)|7.79GB (0.32GB free)|
|VM|100%|

|Item|Value|
|---|---|
|Remote|SSH: git.autodist.com|
|OS|Linux x64 4.19.0-17-amd64|
|CPUs|Intel(R) Xeon(R) Gold 6248 CPU @ 2.50GHz (2 x 2494)|
|Memory (System)|7.79GB (0.32GB free)|
|VM|100%|
</details><details><summary>Extensions (17)</summary>

Extension|Author (truncated)|Version
---|---|---
Bookmarks|ale|13.3.1
jupyter-keymap|ms-|1.0.0
remote-ssh|ms-|0.84.0
remote-ssh-edit|ms-|0.80.0
remote-wsl|ms-|0.66.3
Bookmarks|ale|13.3.1
vscode-intelephense-client|bme|1.8.2
vscode-mysql-client2|cwe|5.6.7
gitlens|eam|12.1.2
php-xdebug|kak|0.0.7
python|ms-|2022.10.1
vscode-pylance|ms-|2022.8.10
jupyter|ms-|2022.6.1201981810
jupyter-keymap|ms-|1.0.0
jupyter-renderers|ms-|1.0.9
php-debug|rob|1.14.0
php-debug|xde|1.27.0

</details><details>
<summary>A/B Experiments</summary>

```
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
vspor879:30202332
vspor708:30202333
vspor363:30204092
vslsvsres303:30308271
pythonvspyl392:30443607
vserr242:30382549
pythontb:30283811
vsjup518:30340749
pythonptprofiler:30281270
vshan820:30294714
vstes263cf:30335440
vscoreces:30445986
pythondataviewer:30285071
vscod805:30301674
binariesv615:30325510
bridge0708:30335490
bridge0723:30353136
vsaa593:30376534
vsc1dst:30438360
pythonvs932:30410667
wslgetstarted:30449410
vscscmwlcmt:30465135
cppdebug:30492333
pylanb8912cf:30529770
vsclangdf:30486550
c4g48928:30535728

```

</details>

**Comment by jrieken** [comment]
@isidorn Can you help me reproduce? With voice over on the mac it works just fine

**Comment by mehgcap** [comment]
I just checked, and the problem is still there. I'm using PHP and Inteliphense, not that that should matter.

I broke a line by removing a required parameter from a constructor call, arrowed up, and hit f8. I landed on the problematic line, which was only spoken once this time. I then went up a couple lines and deleted a blank line. When I did, I heard the line and error again, two or three times over. It doesn't usually happen when reviewing text, at least not now, but it does seem to trigger if any edits are made anywhere in the file.

**Comment by isidorn** [comment]
@mehgcap can you share that full file? So we try to reproduce and understand the issue better. Thank you

**Comment by mehgcap** [comment]
I was just able to reproduce this. I'm on the latest VSCode (public, not insider). Open a new file, with Inteliphense installed and active. Save the new file as a .php file so Inteliphense kicks in, then paste the following:

<?php
function double(int $number): int {
	return $number * 2;
}

double(4);

No problems so far, as this is valid PHP. Now, remove the 4 from the call to the double function. This is an error, as the function requires an argument. You should hear the error sound a second after the 4 is deleted. Up arrow a couple times, then hit f8. You should land on the line calling the function, and be told that you are missing an argument.

So far, so good. Now, arrow up to the line where we define the double function. Hit home, then backspace to remove the blank line just above it. I hear the error again, twice. Arrow up once or twice, and there's no problem, but then arrow back down to the line where double() is defined. Hit end, and then arrow left to the return type. Try deleting it. As I do, I keep getting the error text spoken and brailled, each time twice.

The only way to stop this is to go back down to the call to double() and fix it again. Until I do, I'll get the error text over and over as I edit elsewhere in the file.

**Comment by jvesouza** [comment]
I couldn't find in the comment the text that should be used to reproduce the problem.
I believe it must be a PHP code but I couldn't find it.
Am I doing something wrong?

**Comment by mehgcap** [comment]
No you're not. Looks like Github didn't like me putting code in, so it cut my comment off when I tried to paste it. It also removed all the text I put after my bit of PHP code. I'll try again in a bit.

**Comment by jvesouza** [comment]
I was not able to reproduce the problem using my linux environment, both in the insiders version of vscode and in version 1.73.1.
When I press the f8 key, the error is read by orca, the linux environment screen reader, only once.

**Comment by mehgcap** [comment]
True, I can't always get it to be spoken twice with f8 alone. F8, followed by trying to edit the file as described in my previous comment, is where I get consistent results.

**Comment by isidorn** [comment]
@jvesouza thanks for providing steps. I tried to repro using these steps and I could not. I tried both on Win + nvda and mac + VoiceOver.

Maybe @meganrogge can also try when she finds time.

**Comment by meganrogge** [comment]
I cannot reproduce either - is it possible an extension is causing problems?

**Comment by jvesouza** [comment]
I ran vscode using the --disable-extensions parameter and even then the problem continued to exist.

**Comment by jvesouza** [comment]
@jeanbrazcosta When and if possible, could you try to reproduce this bug by following the steps described in https://github.com/microsoft/vscode/issues/157100#issuecomment-1341343007 ?

**Comment by jeanbrazcosta** [comment]
@jvesouza I was also unable to reproduce the issue in my environment

I have the latest version of NVDA installed and I tried the latest version of insiders but for me the error was read only once.

**Comment by jeanbrazcosta** [comment]
@mehgcap Could it be that you don't have any NVDA addons installed that are causing this?

Have you tried playing with NVDA addons disabled?

**Comment by mehgcap** [comment]
I just tried with add-ons disabled, and nothing changed. I restarted NVDA with add-ons disabled, went back to VSCode, hit f8, then deleted a letter in a comment. The problem that f8 had found was spoken twice and shown on my braille display, just as before.

Could this be down to an object presentation or web setting? I'm not sure which settings would impact this kind of behavior, but if anyone can think of any, I'll try toggling them. Maybe I have something turned on or off that most people have set the other way?

**Comment by mehgcap** [comment]
Now that I think about it, this exact problem happens with language detection. I'd bet money that the cause of one is the cause of the other.

Press ctrl-n to open a new file, then type something that's obviously a specific language. Maybe the opening PHP tag, or DROP TABLE IF EXISTS, or anything else that comes to mind. You want VSCode to detect the language in use.

Once this happens, you'll hear "language [language name] was automatically detected and set as the language mode." For me, almost every time I make any further edits, I hear that same message, usually twice per keystroke. To stop this, I have to save the file, so the language detection goes away.

Both problems behave identically, though it is possible that this one is happening because every edit triggers the language detection again, so the results are being updated and thus read aloud. Does anyone else experience this?

**Comment by joanmarie** [comment]
I can reproduce what @jvesouza described using Orca. It seems to be caused by repeated text-insertion events being emitted by the live region which contains the error. Orca tries to filter such duplicates out, so I'll see if I can give that filtering some more smarts.

**Comment by joanmarie** [comment]
This might prove trickier than I originally thought. I can improve Orca's filtering smarts, but based on the accessibility events I'm seeing, there's no way to distinguish these two scenarios:

1. Press F8, hear the message, press X (duplicate message event should be filtered)
2. Press F8, hear the message, press Escape, then press F8 (message is the same, but not a duplicate)

The reason why is that in both cases, after a key is pressed, the live region is getting destroyed and recreated and firing events.

In scenario 1, when X is pressed, the message remains on screen. Its text is the same. The associated elements presumably are too. So accessibility events indicating the live region was destroyed and recreated is problematic.

In scenario 2, when Escape is pressed, the message goes away. (I've not yet checked to see if the elements are destroyed, hidden, or something else. But conceptually those associated elements have also gone away.) Thus here we would expect accessibility events indicating the live region was destroyed. And when F8 is subsequently pressed to re-display that message, events indicating the live region was recreated/has new text is needed.

@isidorn: Any idea if the element destruction and recreation in scenario 1 is something on your end (as opposed to in Chromium)?

**Comment by isidorn** [comment]
@joanmarie thanks for investigating.

When the screen reader is turned on pressing F8 should pass focus to the inline error widget. @jrieken though I am not seeing that behaviour right now.
Pressing `X` or any other key should not destroy the whole HTML element. If pressing a character affects the error message that will just update the `aria-label` of the HTML element. I just tried and it indeed does that and does not re-create the element.

**Comment by jrieken** [comment]
Can someone with access to NVDA please check if this is still an issue. On mac, with VoiceOver, everything is working as expected

**Comment by mehgcap** [comment]
Yes, it's still happening. Windows 11, NVDA 2024.4, VSCode 1.95.3.

I have a PHP file open. I removed the semicolon from `$i++`, arrowed a couple lines above the newly-created error, and hit f8. I heard the problem, as expected. I hit end, then enter, to get to a new line below the error. I started to type a variable assignment. With each keypress, I heard:

Error. ';' expected.php, $i++, error at 316:9.

I guess it's an improvement that I don't always hear it twice, and the "error" part at the very beginning is also sometimes not there. So while I don't always hear it twice, I still hear it with every character I type.

**Comment by jrieken** [comment]
Please try very latest insiders `1.97.0-insider`

**Comment by mehgcap** [comment]
Is this stable enough for daily use? I use VSCode for work, so I need it to be reliable. I've never tried an Insider build.

**Comment by jrieken** [comment]
It should be pretty stable, at least we, the vscode team, use insiders to develop vscode itself

**Comment by mehgcap** [comment]
Thanks. I also didn't realize it could run alongside the public release.

Using 1.97.0, I can't reproduce the problem after a quick test. It was the same test I ran this morning to confirm that the problem was still present, albeit possibly using different extensions. I had to re-install my extensions for the insider build, so I may have added fewer than are on my public release version. Still, the problem wasn't present.

**Comment by jrieken** [comment]
Ok, that's good news. It might have been resolved with another fix I made last week. Closing this issue but please keep us posted one next stable is related (which will happen early February 2025)
